src/Sentry/Internal/MainExceptionProcessor.cs

```diff
@@ -49,7 +56,23 @@ public void Process(Exception exception, SentryEvent sentryEvent)
 
                 foreach (var key in sentryException.Data.Keys)
                 {
-                    sentryEvent.SetExtra($"Exception[{i}][{key}]", sentryException.Data[key]);
+                    if (key.StartsWith(ExceptionDataTagKey) &&
+                        sentryException.Data[key] is string value &&
+                        key.Length > ExceptionDataTagKey.Length)
+                    {
+                        sentryEvent.SetTag(key.Substring(ExceptionDataTagKey.Length), value);
+                    }
+                    else if (key.StartsWith(ExceptionDataContextKey) &&
+                        sentryException.Data[key] is object contextData &&
+                        key.Length > ExceptionDataContextKey.Length)
+                    {
+                        var keyName = key.Substring(ExceptionDataContextKey.Length);
+                        _ = sentryEvent.Contexts.AddOrUpdate(keyName, contextData, (_, value) => value);
+                    }
+                    else
+                    {
+                        sentryEvent.SetExtra($"Exception[{i}][{key}]", sentryException.Data[key]);
+                    }
```

Comment:
How about first narrowing down by `sentry:` and then by type? Should make the conditions a bit simpler to read.
Also, make sure to use `StringComparison.Ordinal` or maybe even `StringComparison.OrdinalIgnoreCase` to avoid relying on current locale.